### PR TITLE
ns for fx: change dtype cast from once per N node to once per node

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53828 ns for fx: move shadow lstm test to new API
* #53819 ns for fx: move shadow activations linear test to new API
* #53818 ns for fx: move shadow activations conv test to new API
* #53779 ns for fx: add support for lstm activation matching
* #53777 ns for fx: move linear activation test case to new API
* #53776 ns for fx: move compare activations for conv test to new API
* #53772 ns for fx: move lstm dynamic weight test case to new API
* #53765 ns for fx: move linear dynamic weight test case to new API
* #53764 ns for fx: move linear weight test case to new API
* #53761 ns for fx: move conv weight test case to new API
* #53748 ns for fx: extract common code in tests to util functions
* #53729 ns for fx: clean up API naming
* #53249 ns for fx: add support for logging inputs
* **#53248 ns for fx: change dtype cast from once per N node to once per node**
* #53075 ns for fx: change API to take nn.Module instead of GraphModule

Summary:

This PR ensures that when we do a dtype cast for a shadow module,
we insert N dtype casts for N nodes, instead of combining N nodes
into a single dtype cast.

An example where this occurs is `cat([x, y], dim=0)`

```
// original graph

[x, y] -> cat_b -> output

// shadow graph with a single dtype cast, before this PR

  dtype_cast -> cat_a_shadow -> output_a_shadow
  /
[x, y] -> cat_b -> output_b

// shadow graph with multiple dtype casts, after this PR

 [dtype_cast_x, dtype_cast_y] -> cat_a_shadow -> output_a_shadow
 /
[x, y] -> cat_b -> output_b
```

The reason things worked before this PR is because `torch.dequantize`
can take either a single tensor or a list of tensors.  We are changing
this to make an upcoming addition of input loggers easier.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_prepare_model_with_stubs_multiple_dtype_casts
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26931226](https://our.internmc.facebook.com/intern/diff/D26931226)

### CODE REVIEW

**Review by hx89 — APPROVED** [review]

**Comment by vkuzo** [comment]
note: this was merged, not sure why it doesn't show up here
